### PR TITLE
#Task3.1

### PR DESCRIPTION
**Pull request by MEHV29**
This contains the task 3.1

### CODE REVIEW

File: Task3_1/Queue.cs [hunk 1]
```diff
@@ -0,0 +1,107 @@
...
+            return new Queue<T>(_tempQueue);
+        }
+
+        public void Dequeue()
+        {
+            if( _queue.Length == 0 )
+            {
+                throw new Exception("Cannot dequeue from an empty queue");
+            }
+
```
Comment (IgorKheidorov) [inline Task3_1/Queue.cs:81]: Reallocate memory for each operation is not a good idea )) Allocate memory once ( _queue = new T[maximumCapacity];) Create pointers (fileds) to the head and tail of Queue, and shift appropriate field within the same array

File: Task3_1/Queue.cs [hunk 2]
```diff
@@ -0,0 +1,107 @@
...
+using System.Threading.Tasks;
+
+namespace Task3_1
+{
+    internal class Queue<T> : IQueue<T>
+    {
+        private int _size;
+        private int _maximumCapacity;
+        private T[] _queue;
+
```
Comment (IgorKheidorov) [inline Task3_1/Queue.cs:15]: public Queue():  Queue(0) /* use one constructor from other */

File: Task3_1/Queue.cs [hunk 3]
```diff
@@ -0,0 +1,107 @@
...
+        public Queue()
+        {
+            _size = 0;
```
Comment (IgorKheidorov) [inline Task3_1/Queue.cs]: _queue = new T[_maximumCapacit], allocate memory once and reuse it (see next comment)

File: Task3_1/ExtensionMethods.cs [hunk 4]
```diff
@@ -0,0 +1,18 @@
+﻿using System;
+using System.Collections.Generic;
+using System.Linq;
+using System.Text;
+using System.Threading.Tasks;
+
```
Comment (IgorKheidorov) [inline Task3_1/ExtensionMethods.cs:6]: Cool

**Review by IgorKheidorov — CHANGES_REQUESTED** [review]
Redesign the logics-> do not reallocate memory for each operation (too expensive operation, low performance)

File: Task3_1/Queue.cs (hunk 1 above)
Comment (MEHV29) [reply Task3_1/Queue.cs:81]: Modified

File: Task3_1/Queue.cs (hunk 2 above)
Comment (MEHV29) [reply Task3_1/Queue.cs:15]: Modified

File: Task3_1/Queue.cs (hunk 3 above)
Comment (MEHV29) [reply Task3_1/Queue.cs]: Modified

File: Task3_1/ExtensionMethods.cs (hunk 4 above)
Comment (MEHV29) [reply Task3_1/ExtensionMethods.cs:6]: Modified

**Comment by MEHV29** [comment]
I have done the changes, please, let me know anything, thanks!

**Review by IgorKheidorov — APPROVED** [review]
Not absolutely effective but better then previous time and acceptable) Please ask me on the lecture to show the optimum solution))

**Pull request merged**